### PR TITLE
Fix Issue 19754 - cast() sometimes yields lvalue, sometimes yields rvalue

### PR DESCRIPTION
**Pull request by RazvanN7**
Casts were viewed as rvalues. This required some other fixes to previous wrongfully implemented patches.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]
Please add `TEST_OUTPUT`s to those that lack them.

**Comment by UplinkCoder** [comment]
@RazvanN7 Please explain the change in the commit message.
I assume that technically this is a breaking change in semantics, but I'd like to have a explanation of what this code is supposed to do.

**Comment by kinke** [comment]
Yeah, what happens now when casting some `int` lvalue to, say, `float`? Will that be an inacceptable reinterpret-cast with this?

**Comment by UplinkCoder** [comment]
to make the example more concrete.

```
float x = 2.0f;
assert(&x !=&cast(int)x) ??
```

will this pass, fail, or a compile error?

**Comment by RazvanN7** [comment]
>Yeah, what happens now when casting some int lvalue to, say, float? Will that be an inacceptable reinterpret-cast with this?

The compiler will accept code, but only in @system context.

**Comment by UplinkCoder** [comment]
@kinke could you block this from being auto-merged by requesting a chage, I don't want this to go in by accident.

**Comment by UplinkCoder** [comment]
@RazvanN7 what about the addresses?

**Comment by RazvanN7** [comment]
>will this pass, fail, or a compile error?

Compiler error: incompatible types for (& x) !is (&cast(int)x): float* and int*

File: test/compilable/test13336b.d [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+---
+*/
+int sx;
+double sy;
+
+ref f1(bool f)
+{
+    if (f)
+        return sx;
+    return sy;
```
Comment (rainers) [inline test/compilable/test13336b.d]: Why is this supposed to pass now? Other changes to tests look suspicious, too. Can you explain?

**Review by rainers — CHANGES_REQUESTED** [review]

File: test/compilable/test13336b.d (hunk 1 above)
Comment (RazvanN7) [reply test/compilable/test13336b.d]: Without my patch, if you have this code:

```d
int sx;
double sy;

double f1(bool f)
{
     if (f)
        return sx;
     return sy;
}
```

The code compiles. If we add `ref` to the return statement it doesn't compile anymore stating that "`cast(double)sx` is not an lvalue". Why wouldn't the cast be an lvalue? If we want to add other rules that prevent a specific class of casts then that should be discussed in a different context.

Comment (RazvanN7) [reply test/compilable/test13336b.d]: We can discuss other changes in tests that look "suspicious" on a case by case basis.

Comment (rainers) [reply test/compilable/test13336b.d]: > Why wouldn't the cast be an lvalue?

You cannot return a reference to both an integer and a double at the same time.

**Comment by kinke** [comment]
> The compiler will accept the code, but only in @System context.

Absolutely unacceptable. Casting an `int` lvalue to `float` has to yield a new, converted rvalue. `bool CastExp.isLvalue() { return e1.isLvalue(); }` cannot work.

File: src/dmd/typesem.d [hunk 2]
```diff
@@ -1354,6 +1355,13 @@ extern(C++) Type typeSemantic(Type t, Loc loc, Scope* sc)
                         e = new AddrExp(e.loc, e);
                         e = e.expressionSemantic(argsc);
                     }
+                    if (isRefOrOut && !MODimplicitConv(e.type.mod, fparam.type.mod))
```
Comment (UplinkCoder) [inline src/dmd/typesem.d]: can be merged with the block above

**Review by Geod24 — CHANGES_REQUESTED** [review]
The premise of the bugzilla report is IMHO wrong, and I answered there.
TL;DR: `cast` only yields an lvalue for casting away `shared`, not for `const`, `immutable` or types.

**Comment by RazvanN7** [comment]
I see. Thanks for the explanations. Closing this.

**Comment by Geod24** [comment]
@RazvanN7 : Some of those changes were good though. The error message when a function returns mismatched type for example is currently leaking DMD's internal, by talking about a `cast` that does not exist in user-code.
IMO you could submit that part independently.
